nilearn/decomposition/base.py

```diff
@@ -0,0 +1,535 @@
+"""
+PCA dimension reduction on single subjects
+"""
+import atexit
+import os
+from math import ceil
+from tempfile import mkstemp
+import warnings
+import nibabel
+
+import numpy as np
+from scipy import linalg
+from sklearn.externals.joblib import Memory, Parallel, delayed
+from sklearn.linear_model import LinearRegression
+from sklearn.utils.extmath import randomized_svd
+from sklearn.base import BaseEstimator
+
+from ..input_data.nifti_masker import filter_and_mask
+from ..input_data import NiftiMapsMasker
+from ..input_data.masker_validation import check_embedded_nifti_masker
+from .._utils.cache_mixin import CacheMixin, cache
+from .._utils.niimg_conversions import check_niimg_4d
+from nilearn.datasets.utils import _get_dataset_dir
+
+# WindowsError only exist on Windows
+try:
+    WindowsError
+except NameError:
+    WindowsError = None
+
+
+def _delete_file(file_descriptor, file_path, warn=False):
+    """Utility function to cleanup a temporary folder if still existing.
+    Copy from joblib.pool (for independance)"""
+    try:
+        os.close(file_descriptor)
+    except OSError:
+        pass
+    if os.path.exists(file_path):
+        folder_path = os.path.dirname(file_path)
+        os.remove(file_path)
+        # if folder_path is empty
+        if not os.listdir(folder_path):
+            try:
+                # This can fail under windows,
+                #  but will succeed when called by atexit
+                os.rmdir(folder_path)
+            except WindowsError:
+                if warn:
+                    warnings.warn("Could not delete temporary folder %s"
+                                  % folder_path)
+
+
+class mask_and_reduce(object):
+    """Mask and reduce provided data with provided masker, using a PCA
+
+    Uses a PCA on time series to reduce data size. For multiple image, the
```

Comment:
perhaps mention somewhere that SVD or RandomizedSVD is used depending on sample/subject ratio...